cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppStateHandler.java

```
@@ -0,0 +1,184 @@
+    applicationLifecycleService.deleteAllStates(appStateRequest);
+    responder.sendStatus(HttpResponseStatus.OK);
+  }
+
+  private void validateInput(String namespace, String appName) throws Exception {
+    NamespaceId namespaceId = validateNamespace(namespace);
+
+    validateApplication(appName, namespaceId);
+  }
+
+  private void validateInput(String namespace, String appName, String stateKey) throws Exception {
+    validateInput(namespace, appName);
+
+    validateStateKey(stateKey);
+  }
+
+  private void validateInput(String namespace, String appName, String stateKey, byte[] stateValue) throws Exception {
+    validateInput(namespace, appName, stateKey);
+
+    validateStateValue(stateValue);
+  }
+
+  private void validateApplication(String appName, NamespaceId namespaceId)
+    throws BadRequestException, ApplicationNotFoundException {
+    if (appName == null) {
+      throw new BadRequestException("Path parameter app-id cannot be empty");
+    }
+
+    ApplicationId appId = namespaceId.app(appName);
+    applicationLifecycleService.validateApplication(appId);
+  }
+
+  private void validateStateKey(String stateKey) throws BadRequestException {
+    if (stateKey == null) {
+      throw new BadRequestException("Path parameter state-key cannot be empty");
+    }
+  }
+
+  private void validateStateValue(byte[] stateValue) throws BadRequestException {
+    if (stateValue == null) {
+      throw new BadRequestException("Path parameter state-value cannot be empty");
+    }
+  }
+
+  private NamespaceId validateNamespace(String namespace) throws Exception {
+    if (namespace == null) {
+      throw new BadRequestException("Path parameter namespace-id cannot be empty");
+    }
+
+    NamespaceId namespaceId = new NamespaceId(namespace);
+    try {
+      if (!namespaceId.equals(NamespaceId.SYSTEM)) {
+        namespaceQueryAdmin.get(namespaceId);
+      }
+    } catch (NamespaceNotFoundException | AccessException e) {
```

don't need to catch and re-throw, can just let it propagate up.